File: src/jsonmlir/operations/dsl.py
"""Constructeurs DSL à arguments positionnels, compatibles basedpyright.

Les classes ``*Op`` restent des modèles Pydantic (validation JSON, codegen).
Ce module expose des fonctions factory typées pour l'écriture manuelle en Python.
"""

from __future__ import annotations

from collections.abc import Sequence

from jsonmlir.operations.base import BaseValue
from jsonmlir.operations.op_alloc import AllocOp
from jsonmlir.operations.op_alloca import AllocaOp
from jsonmlir.operations.op_binary import BinaryOp
from jsonmlir.operations.op_call import CallOp
from jsonmlir.operations.op_cond import CondOp
from jsonmlir.operations.op_constant import ConstOp
from jsonmlir.operations.op_define_function import DefineFunctionOp
from jsonmlir.operations.op_define_struct import DefineStructOp
from jsonmlir.operations.op_function import FunctionOp
from jsonmlir.operations.op_math import MathOp, MathOperator
from jsonmlir.operations.op_module import ModuleJsonOp, ModuleStatement
from jsonmlir.operations.op_operator import OperatorOp
from jsonmlir.operations.op_print import PrintOp
from jsonmlir.operations.op_set import SetOp
from jsonmlir.operations.op_unary import UnaryOp, UnaryOperator
from jsonmlir.operations.op_var import VarOp
from jsonmlir.operations.op_while import WhileOp
from jsonmlir.utils.enum_scalars import Scalar
from jsonmlir.variables.struct_field import StructField
from jsonmlir.variables.ty.ty import TyNode, parse_ty

FieldSpec = tuple[str, str | TyNode, int, int] | StructField


def _parse_ty(value: str | TyNode) -> TyNode:
    if isinstance(value, str):
        return parse_ty(value)
    return value


def _parse_scalar(value: str | Scalar) -> Scalar:
    if isinstance(value, Scalar):
        return value
    return Scalar(value)


def _parse_field(field: FieldSpec) -> StructField:
    if isinstance(field, StructField):
        return field
    name, ty, offset, size = field
    return StructField(name=name, type=_parse_ty(ty), offset=offset, size=size)


def _parse_ope(ope: str | OperatorOp) -> OperatorOp:
    return ope if isinstance(ope, OperatorOp) else OperatorOp(ope)

def _parse_ope_math(ope: str | MathOperator) -> MathOperator:
    return ope if isinstance(ope, MathOperator) else MathOperator(ope)

def _parse_ope_unary(ope: str | UnaryOperator) -> UnaryOperator:
    return ope if isinstance(ope, UnaryOperator) else UnaryOperator(ope)


def Module(body: Sequence[ModuleStatement] = ()) -> ModuleJsonOp:
    return ModuleJsonOp(body=body)


def DefineStruct(
    name: str,
    size: int,
    fields: Sequence[FieldSpec],
) -> DefineStructOp:
    return DefineStructOp(
        name=name,
        size=size,
        fields=[_parse_field(field) for field in fields],
    )


def DefineFunction(
    name: str,
    args: Sequence[tuple[str, TyNode]] = (),
    return_types: Sequence[TyNode] = (),
) -> DefineFunctionOp:
    return DefineFunctionOp(name=name, args=args, return_types=return_types)

def Alloc(
    name: str,
    type: TyNode,
    size: Sequence[int | VarOp] = ()
) -> AllocOp:
    return AllocOp(name=name, type=type, size=size)

def Alloca(
    name: str,
    type: TyNode,
    size: Sequence[int | VarOp] = ()
) -> AllocaOp:
    return AllocaOp(name=name, type=type, size=size)

def Function(
    name: str,
    args: Sequence[tuple[str, TyNode]] = (),
    body: Sequence[BaseValue] = (),
) -> FunctionOp:
    return FunctionOp(name=name, args=args, body=body)


def Var(
    name: str,
    indices: Sequence[int | str | VarOp] = (),
    *,
    type: TyNode | str | None = None,
) -> VarOp:
    return VarOp(
        name=name,
        indices=indices,
        type=_parse_ty(type) if isinstance(type, str) else type,
    )


def Const(
    val: float | int,
    type: str | Scalar = Scalar.i64,
) -> ConstOp:
    return ConstOp(val=val, type=_parse_scalar(type))


def Binary(
    ope: str | OperatorOp,
    lhs: BaseValue,
    rhs: BaseValue,
) -> BinaryOp:
    return BinaryOp(lhs=lhs, rhs=rhs, ope=_parse_ope(ope))


def Unary(
    ope: str | UnaryOperator,
    value: BaseValue,
) -> UnaryOp:
    return UnaryOp(ope=_parse_ope_unary(ope), value=value)


def Set(var: VarOp, val: BinaryOp | ConstOp | VarOp | CallOp | UnaryOp) -> SetOp:
    return SetOp(var=var, val=val)


def While(
    cond: BaseValue,
    thenBlock: Sequence[BaseValue],
) -> WhileOp:
    return WhileOp(cond=cond, thenBlock=thenBlock)


def Cond(
    cond: BaseValue,
    thenBlock: Sequence[BaseValue],
    elseBlock: Sequence[BaseValue] | None = None,
) -> CondOp:
    return CondOp(cond=cond, thenBlock=thenBlock, elseBlock=elseBlock)


def Call(
    name: str,
    args: Sequence[BaseValue] = (),
) -> CallOp:
    return CallOp(name=name, args=args)

def Math(
    ope: str | MathOperator,
    value: BaseValue,
) -> MathOp:
    return MathOp(ope=_parse_ope_math(ope), value=value)


def Print(value: BaseValue) -> PrintOp:
    return PrintOp(value=value)

File: ts-ast/operations/dsl.ts
import { parseScalar, Scalar } from '../utils/enum_scalars.js';
import { FieldType } from '../variables/memory.js';
import type { TyNode } from '../variables/ty/ty_base.js';
import { parseTy, parseTyOrScalar } from '../variables/ty/ty.js';
import type { BaseValue, ModuleStatement } from './base.js';
import { AllocOp, type AllocSize } from './op_alloc.js';
import { AllocaOp } from './op_alloca.js';
import { BinaryOp } from './op_binary.js';
import { CallOp } from './op_call.js';
import { CondOp } from './op_cond.js';
import { ConstOp } from './op_constant.js';
import { DefineFunctionOp } from './op_define_function.js';
import { DefineStructOp } from './op_define_struct.js';
import { FunctionOp, type FunctionArg } from './op_function.js';
import { MathOp } from './op_math.js';
import { ModuleJsonOp } from './op_module.js';
import {
    MathOperator,
    Operator,
    parseMathOperator,
    parseOperator,
    parseUnaryOperator,
    UnaryOperator,
} from './op_operator.js';
import { PrintOp } from './op_print.js';
import { SetOp, type SetValue } from './op_set.js';
import { UnaryOp } from './op_unary.js';
import { VarOp, type VarIndex } from './op_var.js';
import { WhileOp } from './op_while.js';

/** Field spec: ``[name, type, offset, size]`` or an existing ``FieldType``. */
export type FieldSpec =
    | FieldType
    | readonly [string, string | TyNode, number, number];

function parseField(field: FieldSpec): FieldType {
    if (field instanceof FieldType) {
        return field;
    }
    const [name, ty, offset, size] = field;
    return new FieldType(name, parseTyOrScalar(ty), offset, size);
}

function parseArg(arg: readonly [string, string | TyNode]): FunctionArg {
    return [arg[0], parseTyOrScalar(arg[1])];
}

export function Module(body: readonly ModuleStatement[] = []): ModuleJsonOp {
    return new ModuleJsonOp(body);
}

export function DefineStruct(
    name: string,
    size: number,
    fields: readonly FieldSpec[],
): DefineStructOp {
    return new DefineStructOp(name, size, fields.map(parseField));
}

export function DefineFunction(
    name: string,
    args: readonly (readonly [string, string | TyNode])[] = [],
    returnTypes: readonly (string | TyNode)[] = [],
): DefineFunctionOp {
    return new DefineFunctionOp(
        name,
        args.map(parseArg),
        returnTypes.map(t => parseTyOrScalar(t)),
    );
}

export function Function(
    name: string,
    args: readonly (readonly [string, string | TyNode])[] = [],
    body: readonly BaseValue[] = [],
): FunctionOp {
    return new FunctionOp(name, args.map(parseArg), body);
}

export function Var(
    name: string,
    indices: readonly VarIndex[] = [],
    type?: string | TyNode,
): VarOp {
    return new VarOp(
        name,
        indices,
        type === undefined ? undefined : parseTyOrScalar(type),
    );
}

export function Const(
    val: number,
    type: string | Scalar = Scalar.i64,
): ConstOp {
    return new ConstOp(val, parseScalar(type));
}

export function Binary(
    ope: string | Operator,
    lhs: BaseValue,
    rhs: BaseValue,
): BinaryOp {
    return new BinaryOp(parseOperator(ope), lhs, rhs);
}

export function Unary(
    ope: string | UnaryOperator,
    value: BaseValue,
): UnaryOp {
    return new UnaryOp(parseUnaryOperator(ope), value);
}

export function Math(
    ope: string | MathOperator,
    value: BaseValue,
): MathOp {
    return new MathOp(parseMathOperator(ope), value);
}

export function Set(var_: VarOp, val: SetValue): SetOp {
    return new SetOp(var_, val);
}

export function While(
    cond: BaseValue,
    thenBlock: readonly BaseValue[],
): WhileOp {
    return new WhileOp(cond, thenBlock);
}

export function Cond(
    cond: BaseValue,
    thenBlock: readonly BaseValue[],
    elseBlock?: readonly BaseValue[],
): CondOp {
    return new CondOp(cond, thenBlock, elseBlock);
}

export function Call(
    name: string,
    args: readonly BaseValue[] = [],
): CallOp {
    return new CallOp(name, args);
}

export function Print(value: BaseValue): PrintOp {
    return new PrintOp(value);
}

export function Alloc(
    name: string,
    type: string | TyNode,
    size: readonly AllocSize[] = [],
): AllocOp {
    return new AllocOp(name, parseTyOrScalar(type), size);
}

export function Alloca(
    name: string,
    type: string | TyNode,
    size: readonly AllocSize[] = [],
): AllocaOp {
    return new AllocaOp(name, parseTyOrScalar(type), size);
}

// Re-export parseTy for callers that already hold a JSON type blob.
export { parseTy, parseScalar };
